### ISSUE TITLE
Can't get attribute 'Net' on <module '__main__' from 'D:/demo/cnn/test1.py'>

### ISSUE THREAD

**Issue by xiaopengguo1406** [pytorch/pytorch#7812]
I run Pytorch0.4 on the platform of Win10. The training process works well and I saved the model using torch.save(model, net_g_model_out_path); However, it will cause a error "Can't get attribute 'Net' on <module '__main__' from 'D:/demo/cnn/test1.py'>" when I try to load the model through "net=torch.load('checkpoint/multi/netG_model_epoch_3.pth')". 
The train.py is as follows:
from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

import torch.tensor
import torch.backends.cudnn as cudn
from PIL import Image
import os
import scipy.io as sio  
import torch.utils.data as Data
from torch.utils.data import DataLoader

from dataset import multi_focusdata

parser = argparse.ArgumentParser(description='PyTorch face Example')

parser.add_argument('--batchSize', type=int, default=64, metavar='N',
                    help='input batch size for training (default: 64)')
parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
                    help='input batch size for testing (default: 1000)')
parser.add_argument('--epochs', type=int, default=1000, metavar='N',
                    help='number of epochs to train (default: 10)')
parser.add_argument('--lr', type=float, default=0.01, metavar='LR',
                    help='learning rate (default: 0.01)')
parser.add_argument('--momentum', type=float, default=0.5, metavar='M',
                    help='SGD momentum (default: 0.5)')
parser.add_argument('--no-cuda', action='store_true', default=False,
                    help='disables CUDA training')
parser.add_argument('--seed', type=int, default=1, metavar='S',
                    help='random seed (default: 1)')
parser.add_argument('--log-interval', type=int, default=10, metavar='N',
                    help='how many batches to wait before logging training status')
parser.add_argument('--dataset', type=str, default='multi', metavar='N',
                    help='how many batches to wait before logging training status')
opt = parser.parse_args()
use_cuda = not opt.no_cuda and torch.cuda.is_available()
opt.cuda = not opt.no_cuda and torch.cuda.is_available()

root_path = "dataset"
#dataset = multi_focusdata(os.path.join(root_path,'train'))
dataset = multi_focusdata(os.path.join(root_path,'train'))
train_loader = DataLoader(dataset=dataset, batch_size=opt.batchSize, shuffle=True, num_workers=1)
def weights_init(m):
    classname = m.__class__.__name__
    if classname.find('Conv') != -1:
        m.weight.data.normal_(0.0, 0.02)
    elif classname.find('BatchNorm2d') != -1 or classname.find('InstanceNorm2d') != -1:
        m.weight.data.normal_(1.0, 0.02)
        m.bias.data.fill_(0)
        
        
        
def net_module(in_channels,out_channels,kernel_size,stride,padding):
    return nn.Sequential(
            nn.Conv2d(in_channels=in_channels,out_channels=out_channels,kernel_size=kernel_size,stride=stride,padding=padding),
            nn.ReLU()
           )
 

class Net(nn.Module):
    def __init__(self):
        super(Net,self).__init__()
        self.conv1=net_module(6,32,3,1,2)
        self.conv2=net_module(32,64,3,1,2)
        self.conv3=net_module(64,128,3,1,2)
        self.conv3_drop = nn.Dropout2d()
        self.fc1 = nn.Linear(128*3*3,512)#512=2^x
        self.fc2 = nn.Linear(512,2)
    
    def forward(self,x):
        x=F.relu(F.max_pool2d(self.conv1(x),2))
        x=F.relu(F.max_pool2d(self.conv2(x),2))
        x=F.relu(F.max_pool2d(self.conv3_drop(self.conv3(x)),2))
        
        x=x.view(-1,128*3*3)

        x=F.relu(self.fc1(x))

        x=F.dropout(x,training=self.training)
        x=self.fc2(x)##?
      
        return F.log_softmax(x,dim=1)
    
device = torch.device("cuda")
loss_func=nn.NLLLoss()
img = torch.FloatTensor(opt.batchSize,3, 256, 256)

gt=torch.LongTensor(opt.batchSize,1, 256, 256)
model=Net().to(device)
model.apply(weights_init)

img=img.to(device)
gt=gt.to(device)
loss_func=loss_func.to(device)
optimizer =torch.optim.SGD(model.parameters(), lr=0.001, momentum=0.9)

def train(epoch):
    
    model.train()
    for step,(img_cat,gt1)in enumerate(train_loader):
        img.resize_(img_cat.size()).copy_(img_cat)
        gt.resize_(gt1.size()).copy_(gt1)
        
        optimizer.zero_grad() 
        output=model(img)
        loss=loss_func(output,gt.view(-1))
        loss.backward()
        optimizer.step()
        if step % opt.log_interval == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format( epoch, step * len(img_cat), len(train_loader.dataset),
                100. * step / len(train_loader), loss.data[0]))
            
def checkpoint(epoch):
    if not os.path.exists("checkpoint"):os.mkdir("checkpoint")
    if not os.path.exists(os.path.join("checkpoint", opt.dataset)):
        os.mkdir(os.path.join("checkpoint", opt.dataset))
    net_g_model_out_path = "checkpoint/{}/netG_model_epoch_{}.pth".format(opt.dataset, epoch)

    torch.save(model, net_g_model_out_path)
    print("Checkpoint saved to {}".format("checkpoint/" + opt.dataset))           

if __name__ == '__main__':
    for epoch in range(1, opt.epochs + 1):
        train(epoch)
        #test()
        if epoch % 1 == 0:
            checkpoint(epoch)

When I run the "net=torch.load('checkpoint/multi/netG_model_epoch_3.pth')" error happens.

I am very confused with it! Could you give me a idear how to debug it? Thank you very much!

**Comment by xiaopengguo1406** [comment]
为什么这种存储和加载方式有时候可以工作，有时候会报错呢?是因为0.4版本的pytorch不稳定吗？

**Comment by SsnL** [comment]
This is how Python serialization works, not a PyTorch issue. 

Since you saved a custom object, when it loads it, it needs to use the class definition to reconstruct the object, which will cause issue if it can't find the class definition.

I'm replying in English since this is an English GitHub and people can refer to this in future.

**Comment by xiaopengguo1406** [comment]
With your explanation and the toturial you recommended, I've got it. Thank you very much, have a nice day!

**Comment by LMC4S** [comment]
> 为什么这种存储和加载方式有时候可以工作，有时候会报错呢?是因为0.4版本的pytorch不稳定吗？

Why all the harsh thumb down against this post. Is it wrong to propose questions in one's native language? Cmon

**Comment by LMC4S** [comment]
> 为什么这种存储和加载方式有时候可以工作，有时候会报错呢?是因为0.4版本的pytorch不稳定吗？

Why all the harsh thumb down against this post? Will it make u feel better if it was asked in French?
